### ISSUE TITLE
Mahalanobis nans when not scaling data: RuntimeWarning: invalid value encountered in sqrt

### ISSUE THREAD

**Issue by sergiocalde94** [mosegui/mahalanobis#1]
Hi,

First of all thanks for your package 🙌 .

I'm trying to compute Mahalanobis distance for some datasets from A/B experiments and sometimes I'm encountering a problem that is already solved scaling the input data, but this is not logical to me because as far as I understood about Mahalanobis computation it's not necessary to scale the data.

For example, this occurs for this synthetical example, let's first generate the data:

```python
import pandas as pd
import numpy as np

from mahalanobis import Mahalanobis
from sklearn.preprocessing import StandardScaler

np.random.seed(42)


N_FEATURES = 10
SIZE = 100_000

df_treatment = pd.DataFrame(
    np.random.multinomial(10, [1 / N_FEATURES] * N_FEATURES, size=SIZE),
    columns=[f'feature_{n_feature}' for n_feature in range(N_FEATURES)]
)

df_different = pd.DataFrame(
    np.random.multinomial(10, [1 / 2 ** i for i in range(1, N_FEATURES + 1)], size=SIZE),
    columns=[f'feature_{n_feature}' for n_feature in range(N_FEATURES)]
)

df_same = pd.DataFrame(
    np.random.multinomial(10, [1 / N_FEATURES] * N_FEATURES, size=SIZE),
    columns=[f'feature_{n_feature}' for n_feature in range(N_FEATURES)]
)

df_control = (
    pd
    .concat([df_different, df_same])
    .reset_index(drop=True)
)
```

Now let's check the average distance from treatment to control:

```python
mahalanobis = Mahalanobis(df_control.values, len(df_control))

mahalanobis_distances = mahalanobis.calc_distances(
    df_treatment.values
)
mahalanobis_distances.mean()
```

This piece of code throws:

`RuntimeWarning: invalid value encountered in sqrt
  list(map(lambda difference: np.sqrt(np.dot(np.dot(difference, self._inv_cov_matrix), difference.reshape(-1, 1))), diff_array)))`

And the avegare result is: **nan**.

But if we apply standard scaling the warning disappears and the average result is: **3.48**

```python
sc = StandardScaler()

df_control_scaled = sc.fit_transform(df_control)
df_treatment_scaled = sc.transform(df_treatment)

mahalanobis = Mahalanobis(df_control_scaled, len(df_control_scaled))

mahalanobis_distances = mahalanobis.calc_distances(
    df_treatment_scaled
)
mahalanobis_distances.mean()
```

Any ideas about why is this happening? I run a lot of simulations comparing scaled and non-scaled datasets and the mean distances are always almost the same, so maybe a hotfix is to standardize the dataset, but I'm not super confident about this and I don't understand why this is happening.

Thank in advance! 👍 

**`requirements.txt`**:

```txt
pandas==1.1.3
mahalanobis==1.2.0
numpy==1.20.0
scikit_learn==0.24.2
```

**Comment by mosegui** [comment]
@sergiocalde94 Thanks for this issue. I will take a look at it in the next days (likely not before the weekend). Feel free to PR if you find the issue in the meantime. Best D.

**Comment by mosegui** [comment]
I have been looking over this issue. I have been able to reproduce it and there are a few things I noticed:

The problem seems to have to do with the data set you generated. I will show you different small modifications you can do to the data set in the above example so that this issue does not appear.

With the data sets you provided, the multidimensional mean for the calibration data can be calculated without problems, but when calculating the distances for a new data set (*which is statistically equal to half of the data used for calibration*), negative numbers appear when multiplying `(data - mean).T x (cov_matriv)⁻¹ x (data - mean)`. Pulling the squared root of those negative products yield some NaN entries in the distances array that spoil the mean you attempt to calculate.

I doubt the data set suggested is sound with regard to the attempted calculations. Here are some things you can do so that this problem vanishes (none of them have to do with the `mahalanobis` package).

- Generate a different number of samples: Instead of `SIZE = 100000` try `SIZE = 10000` or `SIZE = 1000000`.
- Use a different number of features: Instead of `N_FEATURES = 10` try `N_FEATURES = 11` or `N_FEATURES = 9`.
- *DO NOT* generate `df_treatment` before `df_same` and `df_different`, i.e. try the following:
```
import pandas as pd
import numpy as np
from mahalanobis import Mahalanobis

np.random.seed(42)


N_FEATURES = 10
SIZE = 100000

df_same = pd.DataFrame(np.random.multinomial(10, [1 / N_FEATURES] * N_FEATURES, size=SIZE),
                       columns=[f'feature_{n_feature}' for n_feature in range(N_FEATURES)]
                       )

df_treatment = pd.DataFrame(np.random.multinomial(10, [1 / N_FEATURES] * N_FEATURES, size=SIZE),
                            columns=[f'feature_{n_feature}' for n_feature in range(N_FEATURES)]
                            )

df_different = pd.DataFrame(np.random.multinomial(10, [1 / 2 ** i for i in range(1, N_FEATURES + 1)][::-1], size=SIZE),
                            columns=[f'feature_{n_feature}' for n_feature in range(N_FEATURES)]
                            )

df_control = (pd.concat([df_different, df_treatment]).reset_index(drop=True))

mahalanobis = Mahalanobis(df_control.values, len(df_control))
mahalanobis_distances = mahalanobis.calc_distances(df_same.values)
```
or 

```
import pandas as pd
import numpy as np
from mahalanobis import Mahalanobis

np.random.seed(42)


N_FEATURES = 10
SIZE = 100000

df_same = pd.DataFrame(np.random.multinomial(10, [1 / N_FEATURES] * N_FEATURES, size=SIZE),
                       columns=[f'feature_{n_feature}' for n_feature in range(N_FEATURES)]
                       )

df_different = pd.DataFrame(np.random.multinomial(10, [1 / 2 ** i for i in range(1, N_FEATURES + 1)][::-1], size=SIZE),
                            columns=[f'feature_{n_feature}' for n_feature in range(N_FEATURES)]
                            )

df_treatment = pd.DataFrame(np.random.multinomial(10, [1 / N_FEATURES] * N_FEATURES, size=SIZE),
                            columns=[f'feature_{n_feature}' for n_feature in range(N_FEATURES)]
                            )

df_control = (pd.concat([df_different, df_treatment]).reset_index(drop=True))

mahalanobis = Mahalanobis(df_control.values, len(df_control))
mahalanobis_distances = mahalanobis.calc_distances(df_same.values)
```

- *DO NOT* reseed the MT19937. The reseeding method `np.random.seed` is there for legacy reasons, and even according to its very documentation it is not recommended reseeding the generator (https://numpy.org/doc/stable/reference/random/generated/numpy.random.seed.html). Seeding a random number generator is done mostly for debugging purposes. If you want truly random numbers, do not seed it. In particular, if your comment out the line `np.random.seed(42)` the problem vanishes.

*TL;DR*
My understanding as of now is that either there is some issue in the random numbers you are generating for this example, although it is also not clear to me that the fact that half of the calibration data is statistically identical to the test data might not be causing some pathology in the calculated covariance matrix. I reach this conclusion through the fact that there are several ways to generate the data so that this strange behavior does not appear, as discussed above.

In particular, I am most inclined to doubt of the randomness of the data due to the fact that the problem can be solved by not seeding the generator or generating the calibration/test data sets in different order.

I am going to close this issue, but feel free to open a new one if you are able to put together a better example that incurs in this problem. With smaller data sets and where the randomness of the data points is examined for soundness beforehand.

Best,

D.
